[RFC] arch: nrf: Panic when building for the wrong board

SebastianBoe opened the pull request:
There are no safety mechanisms to give the user feedback that he might
have choosen a BOARD that doesn't match the connected board. Running
FW for one board on another can lead to obscure debugging.

This patch uses 16 bytes of code to block the application from running
when it is detected that the user has selected the wrong board. After
this patch the user can use his debugger to see why his application is
hanging and it will be obvious what the user error is.

If this method is OK'd I'll port this to nrf51.

Well thought out suggestions for better ways to solve the same usability 
problem are welcome.

PS: I'll make an effort to get this upstreamed as well.

SebastianBoe commented:
@ioannisg: Ping.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/soc.c:
```diff
@@ -373,6 +375,34 @@ static void clock_init(void)
+        // up in the stack trace when the FW is debugged.
+
+    wrong_board_block_forever:
+        goto wrong_board_block_forever;
```
call k_panic() instead

SebastianBoe replied:
I assumed this code would be run before k_panic() could be run, will double-check that assumption.

SebastianBoe replied:
Fixed.

k_panic uses "kernel services" like the uart driver to be able to do things like printk(),
system_init is run before these kernel services are ready (PRE_KERNEL_1), so k_panic
was misbehaving (no printk's) when it was invoked here.

So I moved it to a later point (POST_KERNEL) in the boot process and created a file soc_post_kernel.c in case any other Soc-code needs kernel services in the future.

This reorganization unfortunately increased the code size from 16 bytes to 40,
but I guess we can live with that.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/soc.c:
```diff
@@ -373,6 +375,34 @@ static void clock_init(void)
+
+    if(soc == 0xb0e)
+    {
+        // We are unable to guess what SoC we have built this FW for,
```
No C++-style comments

SebastianBoe replied:
Done.

carlescufi commented on arch/arm/soc/nordic_nrf5/nrf52/soc.c:
```diff
@@ -373,6 +375,34 @@ static void clock_init(void)
 	SystemCoreClock = __SYSTEM_CLOCK_64M;
 }
 
+static void block_forever_if_built_for_wrong_chip(void)
```
rename this to ic_check() or similar

SebastianBoe replied:
Done

SebastianBoe committed:
arch: nrf: Block forever when building for the wrong board

There are no safety mechanisms to give the user feedback that he might
have choosen a BOARD that doesn't match the connected board. Running
FW for one board on another can lead to obscure debugging.

This patch uses 40 bytes of code to block the application from running
when it is detected that the user has selected the wrong board. After
this patch the user can use his debugger to see why his application is
hanging and it will be obvious what the user error is.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
This is ugly, will fix this on the runner side later.
